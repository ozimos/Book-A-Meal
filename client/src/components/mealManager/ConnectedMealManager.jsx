import React from 'react';
import ReactModal from 'react-modal';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Dropzone from 'react-dropzone';
import SearchInput, { createFilter } from 'react-search-input';
import { push } from 'react-router-redux';
import Formsy from 'formsy-react';
import MyFormsyInput from '../helpers/MyInput';
import ConnectedMealOptionsCard from '../mealCard/ConnectedMealOptionsCard';
import MyFormsyTextArea from '../helpers/MyTextArea';
import ProgressLoader from '../helpers/ProgressLoader';
import MealCardContainer from '../mealCard/MealCardContainer';
import Greeting from '../greeting/Greeting';
import { mealActions } from '../../redux/actions';
import ConnectedNav from '../nav/ConnectedNav';
import imageUpload from '../../services/imageUpload';
import '../../../public/styles/search-input.css';
import '../../../public/styles/bookameal.scss';
import '../../../public/styles/modalOpenButton.scss';

ReactModal.setAppElement(document.getElementById('root'));
export class MealManager extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      canSubmit: false,
      showNewMealModal: false,
      displayImage: '',
      searchTerm: '',
      uploading: false,
      uploadPercent: 0
    };
    this.handleImageDrop = this.handleImageDrop.bind(this);
    this.setUploadPercent = this.setUploadPercent.bind(this);
  }
  componentDidMount() {
    if (!this.props.authenticated) {
      this.props.dispatch(push('/login'));
    }
    this.props.dispatch(mealActions.getAllUserMeals());
  }
  setUploadPercent(percentProgress) {
    this.setState({ uploadPercent: percentProgress });
  }
  handleImageDrop(files) {
    this.setState({ uploading: true });
    imageUpload(files, this.setUploadPercent).then((response) => {
      const { data } = response;
      const fileURL = data.secure_url;
      this.urlInput.props.setValue(fileURL);
      this.setState({
        displayImage: fileURL, uploading: false, uploadPercent: 0
      });

    });
  }
  openNewMealModal = () =>
    this.setState({ showNewMealModal: true });
  closeNewMealModal = () =>
    this.setState({ showNewMealModal: false, displayImage: '' });
  handleSubmit = async (meal) => {
    await this.props.dispatch(mealActions.createMeal(meal));
    this.setState({ displayImage: '' });
    this.closeNewMealModal();
  }
  disableButton = () =>
    this.setState({ canSubmit: false });
  enableButton = () =>
    this.setState({ canSubmit: true });
  serverFeedback = error =>
    this.formEl.updateInputsWithError(error);
  searchUpdated = (term) => {
    this.setState({ searchTerm: term });
  }

  render() {
    const KEYS_TO_FILTERS = ['id', 'title', 'description', 'price'];
    const filteredMeals = this.props.meals
      .filter(createFilter(this.state.searchTerm, KEYS_TO_FILTERS));
    const { isCaterer, firstName } = this.props.user;
    return (
      <div className="contain">
        <header className="header">
          <ConnectedNav />
        </header>
        <main>
          <Greeting isCaterer={isCaterer} firstName={firstName} />
          <div className="title-element flexbox">
            <h5 className="shrink">
              Your Meals
            </h5>

            <button
              className="btn title-button"
              onClick={this.openNewMealModal}
            >
              <p>Add Meal</p>
            </button>
          </div>
          <SearchInput
            className="search-input input-field"
            onChange={this.searchUpdated}
          />

          <MealCardContainer
            meals={filteredMeals}
            MealCard={ConnectedMealOptionsCard}
            addClass="scroll2"
            connecting={this.props.connecting}
          />
        </main>
        <ReactModal
          isOpen={this.state.showNewMealModal}
          contentLabel="Input Modal"
          className="modal-content"
          onRequestClose={this.closeNewMealModal}
          shouldCloseOnOverlayClick
        >
          <div className="title flexbox navbar-fixed">
            <h4 className="shrink">
              Meal Editor
            </h4>
            <button
              className="btn title-button"
              onClick={this.closeNewMealModal}
            >
                &#10006;
            </button>
          </div>
          <div className="form3-grid">
            <div>
              <Formsy
                className="form3"
                onValidSubmit={this.handleSubmit}
                onValid={this.enableButton}
                onInvalid={this.disableButton}
                ref={(form) => { this.formEl = form; }}
              >
                <MyFormsyInput
                  typeOfInput="text"
                  name="title"
                  required
                  placeholder="Meal Title"
                  validations={{ minLength: 1, maxLength: 48 }}
                  validationError="Please enter the meal title"
                  validationErrors={{
                minLength: 'input must be longer than 1 character',
                maxLength: 'input must be shorter than 50 characters',
              }}
                />
                <MyFormsyInput
                  typeOfInput="number"
                  name="price"
                  required
                  placeholder="Price"
                  validations={{
                  isOnlyInt: (values, value = '') =>
                  /^(\s?|[1-9]\d*)$/.test(value),
                  minLength: 1,
                  maxLength: 48
                }}
                  validationError="Please enter a price"
                  validationErrors={{
                isOnlyInt: 'price must be integer',
                minLength: 'input must be longer than 1 character',
                maxLength: 'input must be shorter than 50 characters',
              }}
                />
                <MyFormsyTextArea
                  name="description"
                  required
                  placeholder="Description"
                  validations={{ minLength: 5, maxLength: 98 }}
                  validationErrors={{
                minLength: 'input must be longer than 5 characters',
                maxLength: 'input must be shorter than 100 characters',
              }}
                />

                <MyFormsyInput
                  ref={(urlInput) => { this.urlInputMain = urlInput; }}
                  innerRef={(c) => { this.urlInput = c; }}
                  style={{ display: 'none' }}
                  required
                  typeOfInput="url"
                  name="imageUrl"
                  validations={{ isUrl: true, minLength: 5, maxLength: 98 }}
                  validationError="Please select an image"
                  validationErrors={{
                isUrl: 'A valid url was not supplied',
                minLength: 'input must be longer than 5 characters',
                maxLength: 'input must be shorter than 100 characters',
              }}
                />
              </Formsy>
            </div>
            <div className="overlay-container">
              <div className="overlay full" >
                <Dropzone
                  onDrop={this.handleImageDrop}
                  multiple
                  accept="image/*"
                  className="dropzone"
                >
                  {!this.state.uploading &&
                  <button className="btn"> Select an Image</button>}
                </Dropzone>

              </div>

              <div id="meal_image">
                {this.state.displayImage ?
                 !this.state.uploading &&
                 <img
                   src={this.state.displayImage}
                   alt="meal"
                   className="fluid-img"
                 />
              : !this.state.uploading &&
                <div
                  style={{ display: 'flex', 'align-items': 'center' }}
                  className="dropzone"
                >
                  <button className="btn"> Select an Image</button>
                </div>}
                {this.state.uploading &&
                <ProgressLoader upload={this.state.uploadPercent} />
                  }
              </div>
            </div>

            <button
              className={(this.state.canSubmit || !this.props.connecting)
              ? 'btn' : 'btn btn-disabled'}
              onClick={() => this.formEl.submit()}
              type="submit"
              disabled={!this.state.canSubmit || this.props.connecting}
            >
              <p>Continue</p>
            </button>
          </div>

        </ReactModal>
      </div>
    );
  }
}
MealManager.defaultProps = {
  authenticated: false,
  connecting: false,
};
MealManager.propTypes = {
  dispatch: PropTypes.func.isRequired,
  authenticated: PropTypes.bool,
  connecting: PropTypes.bool,
  user: PropTypes.shape({
    isCaterer: PropTypes.bool,
    firstName: PropTypes.string
  }).isRequired,
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
};
const mapStateToProps = state =>
  ({ authenticated: state.loginReducer.authenticated,
    user: state.loginReducer.user.data,
    mealError: state.mealsReducer.mealError,
    connecting: state.mealsReducer.connecting,
    meals: state.mealsReducer.meals, });

export default connect(mapStateToProps)(MealManager);
