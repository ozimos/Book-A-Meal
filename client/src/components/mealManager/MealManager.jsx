import React from 'react';
import ReactModal from 'react-modal';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Dropzone from 'react-dropzone';
import SearchInput, { createFilter } from 'react-search-input';
import { push } from 'react-router-redux';
import Formsy from 'formsy-react';
import MyFormsyInput from '../helpers/MyInput';
import MealCard from '../mealCard/MealCard';
import MyFormsyTextArea from '../helpers/MyTextArea';
import ProgressLoader from '../helpers/ProgressLoader';
import MealCardContainer from '../mealCard/MealCardContainer';
import ConnectedGreeting from '../greeting/Greeting';
import { mealActions } from '../../redux/actions';
import ConnectedNav from '../nav/Nav';
import imageUpload from '../../services/imageUpload';
import '../../../public/styles/search-input.css';
import '../../../public/styles/bookameal.scss';
import '../../../public/styles/modalOpenButton.scss';

ReactModal.setAppElement(document.getElementById('root'));
export class MealManager extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      canSubmit: false,
      showModal: false,
      displayImage: '',
      searchTerm: '',
      uploading: false,
      uploadPercent: 0
    };
    this.handleDrop = this.handleDrop.bind(this);
    this.setUploadPercent = this.setUploadPercent.bind(this);
  }
  componentDidMount() {
    if (!this.props.authenticated) {
      this.props.dispatch(push('/login'));
    }
    this.props.dispatch(mealActions.getAllMeals());
  }
  setUploadPercent(percentProgress) {
    this.setState({ uploadPercent: percentProgress });
  }
  handleDrop(files) {
    this.setState({ uploading: true });
    imageUpload(files, this.setUploadPercent).then((response) => {
      const { data } = response;
      const fileURL = data.secure_url;
      this.urlInput.props.setValue(fileURL);
      this.setState({ displayImage: fileURL, uploading: false, uploadPercent: 0 });

    });
  }
  handleOpenModal = () =>
    this.setState({ showModal: true });
  handleCloseModal = () =>
    this.setState({ showModal: false, displayImage: '' });
  handleSubmit = async (meal) => {
    await this.props.dispatch(mealActions.createMeal(meal));
    this.setState({ displayImage: '' });
    this.handleCloseModal();
  }
  disableButton = () =>
    this.setState({ canSubmit: false });
  enableButton = () =>
    this.setState({ canSubmit: true });
  serverFeedback = error =>
    this.formEl.updateInputsWithError(error);
  searchUpdated = (term) => {
    this.setState({ searchTerm: term });
  }

  render() {
    const KEYS_TO_FILTERS = ['id', 'title', 'description', 'price'];
    const filteredMeals = this.props.meals
      .filter(createFilter(this.state.searchTerm, KEYS_TO_FILTERS));
    const { isCaterer, firstName } = this.props.user;
    return (
      <div className="contain">
        <header className="header">
          <ConnectedNav />
        </header>
        <main>
          <ConnectedGreeting isCaterer={isCaterer} firstName={firstName} />
          <div className="title flexbox">
            <h5 className="shrink">
              Your Meals
            </h5>
            <div className="flexbox">
              <p className="info-spacer">Click on a meal title to edit the meal or</p>
              <button className="btn title-button" onClick={this.handleOpenModal}>
                <p>Add Meal</p>
              </button>
            </div>
          </div>
          <SearchInput className="search-input input-field" onChange={this.searchUpdated} />

          <MealCardContainer
            meals={filteredMeals}
            MealCard={MealCard}
            addClass="scroll2"
            connecting={this.props.connecting}
          />
        </main>
        <ReactModal
          isOpen={this.state.showModal}
          contentLabel="Input Modal"
          className="modal-content"
          onRequestClose={this.handleCloseModal}
          shouldCloseOnOverlayClick
        >
          <div className="title flexbox">
            <h4 className="shrink">
              Meal Editor
            </h4>
            <div className="flexbox">
              <button className="btn title-button" onClick={this.handleCloseModal}>
                &#10006;
              </button>
            </div>
          </div>
          <Formsy
            className="form3"
            onValidSubmit={this.handleSubmit}
            onValid={this.enableButton}
            onInvalid={this.disableButton}
            ref={(form) => { this.formEl = form; }}
          >
            <MyFormsyInput
              typeOfInput="text"
              name="title"
              required
              placeholder="Meal Title"
              validations="minLength:1"
              validationError="Please enter the meal title"
            />
            <MyFormsyInput
              typeOfInput="number"
              name="price"
              required
              placeholder="Price"
              validations={{
                  isOnlyInt: (values, value = '') => /^(\s?|[1-9]\d*)$/.test(value)
                }}
              validationError="Please enter a price"
              validationErrors={{ isOnlyInt: 'price must be integer' }}
            />
            <MyFormsyTextArea
              name="description"
              required
              placeholder="Description"
            />

            <MyFormsyInput
              ref={(urlInput) => { this.urlInputMain = urlInput; }}
              innerRef={(c) => { this.urlInput = c; }}
              style={{ display: 'none' }}
              required
              typeOfInput="url"
              name="imageUrl"
              validations="minLength:5"
              validationError="Please select an image"
            />
          </Formsy>
          <Dropzone
            onDrop={this.handleDrop}
            multiple
            accept="image/*"
            className="dropzone"
          >
            {!this.state.uploading &&
            <button className="btn"> Select an Image</button>}
          </Dropzone>
          {this.state.uploading &&
          <ProgressLoader upload={this.state.uploadPercent} />
                  }
          <div id="meal_image">
            {this.state.displayImage ? <img src={this.state.displayImage} alt="meal" className="fluid-img" /> : false}
          </div>
          <button
            className={(this.state.canSubmit || !this.props.connecting) ? 'btn' : 'btn btn-disabled'}
            onClick={() => this.formEl.submit()}
            type="submit"
            disabled={!this.state.canSubmit || this.props.connecting}
          >
            <p>Continue</p>
          </button>
        </ReactModal>
      </div>
    );
  }
}
MealManager.defaultProps = {
  authenticated: false,
  connecting: false,
};
MealManager.propTypes = {
  dispatch: PropTypes.func.isRequired,
  authenticated: PropTypes.bool,
  connecting: PropTypes.bool,
  user: PropTypes.shape({
    isCaterer: PropTypes.bool,
    firstName: PropTypes.string
  }).isRequired,
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
};
const mapStateToProps = state =>
  ({ authenticated: state.loginReducer.authenticated,
    user: state.loginReducer.user.data,
    mealError: state.mealsReducer.mealError,
    connecting: state.mealsReducer.connecting,
    meals: state.mealsReducer.meals, });

export default connect(mapStateToProps)(MealManager);
