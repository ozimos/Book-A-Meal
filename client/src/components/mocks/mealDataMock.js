export const rows = [
  { id: 'abc', price: 1200 }, { id: 'def', price: 2000 }
];

export const allMeals = {
  data: { rows }

};

export const meal = {
  data: { id: 'mealId',
    price: 200,
    description: 'nice',
    imageUrl: 'nice',
    title: 'nice'
  }
};
