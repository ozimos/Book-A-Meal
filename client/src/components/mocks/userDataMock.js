export const userData = {
  data: { id: '123', email: 'asdf' },
  message: 'abc',
  token: 'abc'
};
export const userInput = {
  data: { email: 'asdf' }
};
export const userIncludes = {
  email: 'asdf',
  firstName: 'first',
  lastName: 'last'
};
