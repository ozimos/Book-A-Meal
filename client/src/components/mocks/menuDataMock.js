import { allMeals } from './mealDataMock';

export default {
  data: { Meals: allMeals.data }
};