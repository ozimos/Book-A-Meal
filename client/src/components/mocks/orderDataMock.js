import { allMeals } from './mealDataMock';

const Meals = allMeals.data.map(meal => ({ ...meal, MealOrders: { quantity: 2 } }));
export const order = {
  data: { id: 'orderId',
    userId: 'userId',
    Meals,
    mealList: ['abc', 'def'],
    quantityList: [2, 2] }
};
export const allOrders = {
  data: [{ ...order.data, id: 'order1' }, { ...order.data, id: 'order2' }]
};