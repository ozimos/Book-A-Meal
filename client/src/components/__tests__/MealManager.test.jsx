/* global React:false, shallow:false toJson:false */
import { MealManager } from '../mealManager/ConnectedMealManager';
import { allMeals } from '../mocks/mealDataMock';

describe('MealManager Component', () => {
  const user = { isCaterer: true, firstName: 'user' };
  it('renders correctly', () => {
    const wrapper =
    shallow(<MealManager
      dispatch={jest.fn()}
      user={user}
      meals={allMeals.data}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
