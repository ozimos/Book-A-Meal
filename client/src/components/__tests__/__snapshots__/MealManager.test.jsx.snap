// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealManager Component renders correctly 1`] = `
<div
  className="contain"
>
  <header
    className="header"
  >
    <Connect(Nav) />
  </header>
  <main>
    <Greeting
      firstName="user"
      isCaterer={true}
    />
    <div
      className="title flexbox"
    >
      <h4
        className="shrink"
      >
        Your Meals
      </h4>
      <div
        className="flexbox"
      >
        <p
          className="info-spacer"
        >
          Click on a meal title to edit the meal or
        </p>
        <button
          className="btn title-button"
          onClick={[Function]}
        >
          <p>
            Add Meal
          </p>
        </button>
      </div>
    </div>
    <Search
      caseSensitive={false}
      className="search-input input-field"
      fuzzy={false}
      onChange={[Function]}
      throttle={200}
    />
    <MealCardContainer
      MealCard={[Function]}
      addClass=""
      meals={
        Array [
          Object {
            "id": "abc",
            "price": 1200,
          },
          Object {
            "id": "def",
            "price": 2000,
          },
        ]
      }
    />
  </main>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal-content"
    closeTimeoutMS={0}
    contentLabel="Input Modal"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      className="title flexbox"
    >
      <h4
        className="shrink"
      >
        Meal Editor
      </h4>
      <div
        className="flexbox"
      >
        <button
          className="btn title-button"
          onClick={[Function]}
        >
          âœ–
        </button>
      </div>
    </div>
    <Formsy
      className="form3"
      disabled={false}
      getErrorMessage={[Function]}
      getErrorMessages={[Function]}
      getValue={[Function]}
      hasValue={[Function]}
      isFormDisabled={[Function]}
      isFormSubmitted={[Function]}
      isPristine={[Function]}
      isRequired={[Function]}
      isValid={[Function]}
      isValidValue={[Function]}
      mapping={null}
      onChange={[Function]}
      onError={[Function]}
      onInvalid={[Function]}
      onInvalidSubmit={[Function]}
      onReset={[Function]}
      onSubmit={[Function]}
      onValid={[Function]}
      onValidSubmit={[Function]}
      preventExternalInvalidation={false}
      resetValue={[Function]}
      setValidations={[Function]}
      setValue={[Function]}
      showError={[Function]}
      showRequired={[Function]}
      validationErrors={null}
    >
      <Formsy(Input)
        innerRef={[Function]}
        name="title"
        placeholder="Meal Title"
        required={true}
        typeOfInput="text"
        validationError="Please enter the meal title"
        validationErrors={Object {}}
        validations="minLength:1"
      />
      <Formsy(Input)
        innerRef={[Function]}
        name="price"
        placeholder="Price"
        required={true}
        typeOfInput="number"
        validationError="Please enter a price"
        validationErrors={
          Object {
            "isOnlyInt": "price must be integer",
          }
        }
        validations={
          Object {
            "isOnlyInt": [Function],
          }
        }
      />
      <Formsy(TextArea)
        innerRef={[Function]}
        name="description"
        placeholder="Description"
        required={true}
        validationError=""
        validationErrors={Object {}}
        validations={null}
      />
      <Formsy(Input)
        innerRef={[Function]}
        name="imageUrl"
        required={true}
        style={
          Object {
            "display": "none",
          }
        }
        typeOfInput="url"
        validationError="Please select an image"
        validationErrors={Object {}}
        validations="minLength:5"
      />
    </Formsy>
    <t
      accept="image/*"
      className="dropzone"
      disableClick={false}
      disablePreview={false}
      disabled={false}
      maxSize={Infinity}
      minSize={0}
      multiple={true}
      onDrop={[Function]}
      preventDropOnDocument={true}
    >
      <button
        className="btn"
      >
         Select an Image
      </button>
    </t>
    <div
      id="meal_image"
    />
    <button
      className="btn btn-disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      <p>
        Continue
      </p>
    </button>
  </Modal>
</div>
`;
